syncer/sharding-meta/shardmeta.go

```diff
@@ -0,0 +1,237 @@
+	}
+	if isGlobal {
+		meta.global = &ShardingSequence{Items: items}
+	} else {
+		meta.sources[source] = &ShardingSequence{Items: items}
+	}
+	meta.activeIdx = activeIdx
+	return nil
+}
+
+// ActiveIdx returns the activeIdx of sharding meta
+func (meta *ShardingMeta) ActiveIdx() int {
+	return meta.activeIdx
+}
+
+func (meta *ShardingMeta) reinitialize() {
+	meta.activeIdx = 0
+	meta.global = &ShardingSequence{make([]*DDLItem, 0)}
+	meta.sources = make(map[string]*ShardingSequence)
+}
+
+// checkItemExists checks whether DDLItem exists in its source sequence
+// if exists, return the index of DDLItem in source sequence.
+// if not exists, return the next index in source sequence.
+func (meta *ShardingMeta) checkItemExists(item *DDLItem) (int, bool) {
+	source, ok := meta.sources[item.Source]
+	if !ok {
+		return 0, false
+	}
+	for idx, ddlItem := range source.Items {
+		if item.FirstPos.Compare(ddlItem.FirstPos) == 0 {
+			return idx, true
+		}
+	}
+	return len(source.Items), false
+}
+
+// AddItem adds a new comming DDLItem into ShardingMeta
+// 1. if DDLItem already exists in source sequence, check whether it is active DDL only
+// 2. add the DDLItem into its related source sequence
+// 3. if it is a new DDL in global sequence, add it into global sequence
+// 4. check the source sequence is the prefix-sequence of global sequence, if not, return an error
+// returns:
+//   active: whether the DDL will be processed in this round
+func (meta *ShardingMeta) AddItem(item *DDLItem) (active bool, err error) {
+	index, exists := meta.checkItemExists(item)
+	if exists {
+		return index == meta.activeIdx, nil
+	}
+
+	if source, ok := meta.sources[item.Source]; !ok {
+		meta.sources[item.Source] = &ShardingSequence{Items: []*DDLItem{item}}
+	} else {
+		source.Items = append(source.Items, item)
+	}
+
+	found := false
+	for _, globalItem := range meta.global.Items {
+		if utils.CompareShardingDDLs(item.DDLs, globalItem.DDLs) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		meta.global.Items = append(meta.global.Items, item)
+	}
+
+	global, source := meta.global, meta.sources[item.Source]
+	if !source.IsPrefixSequence(global) {
+		return false, errors.Errorf("wrong sql sequence of source %+v, global is %+v", source.Items, global.Items)
+	}
+
+	return index == meta.activeIdx, nil
+}
+
+// GetActiveDDLItem returns the source table's active DDLItem
+// if in DDL unsynced procedure, the active DDLItem means the syncing DDL
+// if in re-sync procedure, the active DDLItem means the next syncing DDL in DDL syncing sequence, may be nil
+func (meta *ShardingMeta) GetActiveDDLItem(tableSource string) *DDLItem {
+	source, ok := meta.sources[tableSource]
+	if !ok {
+		return nil
+	}
+	if meta.activeIdx < len(source.Items) {
+		return source.Items[meta.activeIdx]
+	}
+	return nil
+}
+
+// InSequenceSharding returns whether in sequence sharding
+func (meta *ShardingMeta) InSequenceSharding() bool {
+	return meta.activeIdx < len(meta.global.Items)
```

Comment:
the `activeIdx` begin with 0? if yes the `meta.activeIdx < len(meta.global.Items)` is always true